File: web/src/lib/doc-comments/anchor.ts
/**
 * Turning a reader's text selection into an anchor the API will store, and a
 * stored anchor back into a range on screen.
 *
 * ## The two coordinate spaces, and why the anchor straddles them
 *
 * The document exists twice: as the markdown the API stores, and as the prose
 * ProseMirror renders from it. `**bold**` is eight characters in one and four in
 * the other, so a position in one is not a position in the other.
 *
 * The stored anchor deliberately uses both:
 *
 * - `start` / `end` are byte offsets into the **markdown**, because that is what
 *   the API documents them to be and what any other client — or the server — has
 *   to be able to read them as.
 * - `exact` / `prefix` / `suffix` are taken from the **rendered** text, because
 *   their job is to find the quote again in what the reader is looking at, and a
 *   quote whose context strings live in a different space than the quote itself
 *   cannot disambiguate anything. In plain prose, which is nearly all of it, the
 *   two spaces are character-for-character identical anyway; the distinction only
 *   bites across inline markup, and there it is the rendered text that still
 *   matches the page.
 *
 * That split is the one deliberate choice in this file. Everything else follows
 * from it: `locateQuoteInSource` crosses from rendered text into markdown
 * (markup-tolerant, because the needle has none and the haystack does), and
 * `locateQuoteInRendered` stays inside rendered space (verbatim, because both
 * sides are the same kind of text).
 *
 * ## When the quote cannot be placed
 *
 * `locateQuoteInSource` returning null is a legitimate outcome, not a bug to
 * paper over. The API's third legal anchor state is "quote, offsets NULL" —
 * orphaned — and writing that down is strictly better than inventing offsets
 * that point at the wrong words. The caller warns and stores the quote.
 */

import type { DocumentCommentAnchor } from "@/types";
import { byteLength, utf16ToByteOffset } from "./offsets";

/** How much neighbouring text is kept either side of the quote. */
export const CONTEXT_LENGTH = 48;

/** Longest quote we will store. The API refuses anything over 2000 bytes. */
export const MAX_QUOTE_LENGTH = 500;

/** Above this, the markup-tolerant scan is skipped — see locateQuoteInSource. */
const TOLERANT_SCAN_MAX_SOURCE = 200_000;

/** A [start, end) span in UTF-16 indices. */
export interface CharSpan {
  start: number;
  end: number;
}

// ---------------------------------------------------------------------------
// Scoring
// ---------------------------------------------------------------------------

/** Length of the longest common suffix of `a` and `b`. */
function commonSuffixLength(a: string, b: string): number {
  let n = 0;
  while (n < a.length && n < b.length && a.charAt(a.length - 1 - n) === b.charAt(b.length - 1 - n)) {
    n++;
  }
  return n;
}

/** Length of the longest common prefix of `a` and `b`. */
function commonPrefixLength(a: string, b: string): number {
  let n = 0;
  while (n < a.length && n < b.length && a.charAt(n) === b.charAt(n)) n++;
  return n;
}

/**
 * How well a candidate occupying [start, end) in `haystack` matches the context
 * the anchor remembers. Higher is better; zero means "the quote is here but
 * nothing around it agrees", which is still a candidate when it is the only one.
 */
function contextScore(
  haystack: string,
  start: number,
  end: number,
  prefix: string,
  suffix: string,
): number {
  const before = haystack.slice(Math.max(0, start - prefix.length), start);
  const after = haystack.slice(end, end + suffix.length);
  return commonSuffixLength(before, prefix) + commonPrefixLength(after, suffix);
}

/**
 * Pick the best of several candidate spans: most agreeing context wins, and a
 * tie is broken by whichever sits nearest the position we expected.
 */
function bestCandidate(
  haystack: string,
  candidates: CharSpan[],
  prefix: string,
  suffix: string,
  hint: number | null,
): CharSpan | null {
  const first = candidates[0];
  if (!first) return null;
  if (candidates.length === 1) return first;

  let best = first;
  let bestScore = -1;
  let bestDistance = Number.POSITIVE_INFINITY;

  for (const candidate of candidates) {
    const score = contextScore(haystack, candidate.start, candidate.end, prefix, suffix);
    const distance = hint === null ? 0 : Math.abs(candidate.start - hint);
    if (score > bestScore || (score === bestScore && distance < bestDistance)) {
      best = candidate;
      bestScore = score;
      bestDistance = distance;
    }
  }
  return best;
}

// ---------------------------------------------------------------------------
// Rendered space -> rendered space (verbatim)
// ---------------------------------------------------------------------------

/** Every verbatim occurrence of `needle` in `haystack`. */
function allOccurrences(haystack: string, needle: string): CharSpan[] {
  const spans: CharSpan[] = [];
  if (!needle) return spans;
  let from = 0;
  for (;;) {
    const at = haystack.indexOf(needle, from);
    if (at === -1) break;
    spans.push({ start: at, end: at + needle.length });
    from = at + 1;
    // A pathological document (the same short quote thousands of times) must
    // not turn one render into a quadratic scan.
    if (spans.length > 500) break;
  }
  return spans;
}

/**
 * Find the anchor's quote in the text currently on screen.
 *
 * `hintFraction` is where the anchor's byte offset says it should be, expressed
 * as a fraction of the document, because the rendered text is a different length
 * than the markdown and the byte offset cannot be used directly. It only breaks
 * ties between equally-well-contextualised repeats of the same phrase.
 *
 * Returns null when the quote is not on screen at all — the document has been
 * edited since, and the thread is shown detached rather than highlighted
 * somewhere plausible-looking and wrong.
 */
export function locateQuoteInRendered(
  rendered: string,
  anchor: Pick<DocumentCommentAnchor, "exact" | "prefix" | "suffix">,
  hintFraction: number | null = null,
): CharSpan | null {
  const exact = anchor.exact;
  if (!exact) return null;

  const candidates = allOccurrences(rendered, exact);
  const hint =
    hintFraction === null ? null : Math.round(hintFraction * rendered.length);
  return bestCandidate(rendered, candidates, anchor.prefix, anchor.suffix, hint);
}

// ---------------------------------------------------------------------------
// Rendered space -> markdown space (markup-tolerant)
// ---------------------------------------------------------------------------

/**
 * Characters the markdown may carry that the rendered text does not. Skipping
 * them is what lets a selection of "the bold word" match `the **bold** word`.
 */
const MARKUP_CHARS = new Set(["*", "_", "`", "~", "\\", "[", "#", ">", "|"]);

/** A link's tail — `](https://example.com "title")` — skipped as one unit. */
const LINK_TAIL = /^\]\([^()\s]*(?:\s+"[^"]*")?\)/;

const WHITESPACE = /\s/;

/**
 * Longest run of markdown-only characters allowed between two characters of the
 * quote. Enough for `**`, a fenced backtick run or a short link tail; short
 * enough that the scan cannot wander across a paragraph and call it a match.
 */
const MAX_SKIP_RUN = 48;

/**
 * Try to match `needle` starting exactly at `from` in `source`, allowing the
 * source to contain markdown syntax the rendered text does not, and allowing
 * whitespace runs to differ (a markdown soft line break renders as one space).
 *
 * Returns the end index in `source`, or -1.
 */
function tolerantMatchAt(source: string, needle: string, from: number): number {
  let i = 0; // into needle
  let j = from; // into source

  while (i < needle.length) {
    if (j >= source.length) return -1;

    const nc = needle.charAt(i);
    const sc = source.charAt(j);

    if (nc === sc) {
      i++;
      j++;
      continue;
    }

    // Whitespace is compared as runs: any run matches any run.
    if (WHITESPACE.test(nc) && WHITESPACE.test(sc)) {
      while (i < needle.length && WHITESPACE.test(needle.charAt(i))) i++;
      while (j < source.length && WHITESPACE.test(source.charAt(j))) j++;
      continue;
    }

    // Skip a run of source-only markdown before giving up on this start.
    let skipped = 0;
    while (j < source.length && skipped < MAX_SKIP_RUN) {
      const tail = LINK_TAIL.exec(source.slice(j));
      if (tail) {
        j += tail[0].length;
        skipped += tail[0].length;
        continue;
      }
      if (!MARKUP_CHARS.has(source.charAt(j))) break;
      j++;
      skipped++;
    }
    if (skipped === 0) return -1;
  }

  return j;
}

/**
 * Locate a rendered-text quote inside the markdown it was rendered from.
 *
 * Verbatim first — that is the case for all plain prose, and it is exact. The
 * markup-tolerant scan is the fallback for a selection that crossed a bold run,
 * a link or a soft line break, where the rendered characters exist in the source
 * with syntax interleaved.
 *
 * `hint` is a rough UTF-16 index into `source`, used only to break ties.
 */
export function locateQuoteInSource(
  source: string,
  exact: string,
  prefix: string,
  suffix: string,
  hint: number | null = null,
): CharSpan | null {
  if (!exact) return null;

  const verbatim = allOccurrences(source, exact);
  if (verbatim.length > 0) {
    return bestCandidate(source, verbatim, prefix, suffix, hint);
  }

  if (source.length > TOLERANT_SCAN_MAX_SOURCE) return null;

  const first = exact.charAt(0);
  const candidates: CharSpan[] = [];
  for (let start = 0; start < source.length; start++) {
    // Only start where the first character could match — either literally, or
    // as the head of a whitespace run.
    const sc = source.charAt(start);
    if (sc !== first && !(WHITESPACE.test(first) && WHITESPACE.test(sc))) continue;
    const end = tolerantMatchAt(source, exact, start);
    if (end !== -1) {
      candidates.push({ start, end });
      if (candidates.length > 200) break;
    }
  }

  return bestCandidate(source, candidates, prefix, suffix, hint);
}

// ---------------------------------------------------------------------------
// Building the anchor
// ---------------------------------------------------------------------------

export interface BuiltAnchor {
  anchor: Omit<DocumentCommentAnchor, "orphaned">;
  /**
   * True when the quote could not be placed in the markdown, so the anchor is
   * being stored without offsets and the comment will come back orphaned. The
   * composer says so rather than letting it be a surprise later.
   */
  unplaceable: boolean;
}

/** Trim a selection's edges without dragging the indices out of step. */
function trimSpan(text: string, span: CharSpan): CharSpan {
  let { start, end } = span;
  while (start < end && WHITESPACE.test(text.charAt(start))) start++;
  while (end > start && WHITESPACE.test(text.charAt(end - 1))) end--;
  return { start, end };
}

/**
 * Build the anchor for a selection of the rendered document.
 *
 * @param source   the document's markdown, exactly as the API holds it
 * @param rendered the flattened text of what is on screen
 * @param span     the selection, in `rendered`'s UTF-16 indices
 */
export function buildAnchorFromSelection(
  source: string,
  rendered: string,
  span: CharSpan,
): BuiltAnchor | null {
  const trimmed = trimSpan(rendered, span);
  if (trimmed.end <= trimmed.start) return null;

  // A quote is an identifier, not a copy of the document. Truncating here rather
  // than letting the API refuse the write at 2000 bytes — which a Cyrillic
  // document would hit at half the characters an English one does.
  const end = Math.min(trimmed.end, trimmed.start + MAX_QUOTE_LENGTH);
  const exact = rendered.slice(trimmed.start, end);
  const prefix = rendered.slice(Math.max(0, trimmed.start - CONTEXT_LENGTH), trimmed.start);
  const suffix = rendered.slice(end, end + CONTEXT_LENGTH);

  // Where in the markdown we expect to land, as a rough proportion — the two
  // strings are different lengths, so this is a tie-breaker and nothing more.
  const hint =
    rendered.length > 0
      ? Math.round((trimmed.start / rendered.length) * source.length)
      : null;

  const inSource = locateQuoteInSource(source, exact, prefix, suffix, hint);

  if (!inSource) {
    return {
      anchor: { exact, prefix, suffix, start: null, end: null },
      unplaceable: true,
    };
  }

  return {
    anchor: {
      exact,
      prefix,
      suffix,
      start: utf16ToByteOffset(source, inSource.start),
      end: utf16ToByteOffset(source, inSource.end),
    },
    unplaceable: false,
  };
}

/**
 * Where the anchor claims to sit, as a fraction of the document, for
 * `locateQuoteInRendered`'s tie-break. Null when the anchor is orphaned.
 */
export function anchorHintFraction(
  source: string,
  anchor: Pick<DocumentCommentAnchor, "start">,
): number | null {
  if (anchor.start === null || anchor.start === undefined) return null;
  const total = byteLength(source);
  if (total <= 0) return null;
  return Math.max(0, Math.min(1, anchor.start / total));
}

File: pkg/mdoc/anchor.go
package mdoc

import (
	"fmt"
	"regexp"
	"strings"
	"unicode"
	"unicode/utf8"
)

// Resolving a quotation to a comment anchor, server-side.
//
// ## Why this is not the caller's job
//
// A document comment's anchor is a pair of byte offsets into the markdown plus
// the quote and its neighbours. The web client computes them from a mouse
// selection, where the offsets fall out of the selection itself. An agent has no
// selection: it has a sentence, and it would have to compute the offsets from the
// text — which is exactly where it gets them wrong. Measured on 2026-08-19: a
// Cyrillic quote sitting at byte 853 is reported at 475 by a naive character
// index, because Go, Postgres and this API count bytes while a character count
// counts characters, and Russian is two bytes a letter. An anchor computed that
// way points confidently at somebody else's sentence.
//
// ## Why it mirrors the frontend rather than inventing its own rules
//
// The resolution rules already exist, in web/src/lib/doc-comments/anchor.ts, and
// two independent implementations of "find this quote" drift. When they drift, an
// agent's comment and a human's comment on the same sentence land in different
// places, and the disagreement is invisible until somebody notices two
// highlights where there should be one. So the scan order (verbatim first, then
// markup-tolerant), the scoring (how much of the remembered context agrees) and
// every constant below are the same as that file's, and
// TestResolveQuote_MatchesFrontendFixture pins them to values produced by
// running the real TypeScript.
//
// ## The one deliberate difference
//
// The frontend, faced with several equally good matches, picks one: it has a
// hint — where the selection actually was — and using it is strictly better than
// refusing to place a highlight the reader just made. The server has no hint, so
// where the frontend tie-breaks, this refuses and says how many matches there
// were. Silently picking the first occurrence of "the API" out of eleven is the
// one failure mode that produces a confidently wrong anchor with nothing to
// notice, and an agent can always send more context.

const (
	// contextLength is how much neighbouring text an anchor keeps either side of
	// the quote. Same value as CONTEXT_LENGTH in anchor.ts, counted in the same
	// units (code points; identical to its UTF-16 units for everything below the
	// astral planes, which is all prose).
	contextLength = 48

	// maxQuoteBytes is the longest quote this will resolve. The API's anchor
	// column refuses more.
	maxQuoteBytes = 2000

	// tolerantScanMaxSource is the size above which the markup-tolerant fallback
	// is skipped — it is O(source x quote) in the worst case. Same as
	// TOLERANT_SCAN_MAX_SOURCE.
	tolerantScanMaxSource = 200_000

	// maxSkipRun is the longest run of markdown-only characters allowed between
	// two characters of the quote: enough for `**`, a backtick run or a short link
	// tail, short enough that the scan cannot wander across a paragraph and call
	// it a match. Same as MAX_SKIP_RUN.
	maxSkipRun = 48

	// maxCandidates and maxTolerantCandidates bound a pathological document — the
	// same phrase thousands of times — from turning one request into a quadratic
	// scan. Same as the caps in allOccurrences and locateQuoteInSource.
	maxCandidates         = 500
	maxTolerantCandidates = 200
)

// markupChars are the characters the markdown may carry that the rendered text
// does not. Skipping them is what lets a quote of "the bold word" match
// `the **bold** word`. Same set as MARKUP_CHARS.
var markupChars = map[byte]bool{
	'*': true, '_': true, '`': true, '~': true, '\\': true,
	'[': true, '#': true, '>': true, '|': true,
}

// linkTail matches a link's tail — `](https://example.com "title")` — which is
// skipped as one unit. Same expression as LINK_TAIL.
var linkTail = regexp.MustCompile(`^\]\([^()\s]*(?:\s+"[^"]*")?\)`)

// Anchor is a resolved comment anchor: a W3C Web Annotation selector pair, the
// shape document_comments stores.
//
// Start and End are BYTE offsets into the markdown, half-open [Start, End) —
// the units the anchor_start/anchor_end columns are documented in. Slicing the
// body with them is the whole point, and it is what the Cyrillic test asserts.
type Anchor struct {
	Exact  string `json:"exact"`
	Prefix string `json:"prefix"`
	Suffix string `json:"suffix"`
	Start  int    `json:"start"`
	End    int    `json:"end"`
}

// QuoteNotFoundError reports a quote that does not occur in the document.
type QuoteNotFoundError struct{ Quote string }

func (e *QuoteNotFoundError) Error() string {
	return "no such quote in this document: the text was not found, either verbatim or with " +
		"markdown syntax between its characters. Quote the document exactly as it reads, " +
		"including punctuation."
}

// AmbiguousQuoteError reports a quote that occurs more than once and could not be
// narrowed to one by the context supplied. Matches is how many times it occurs,
// which is the number that tells a caller whether to add context or pick a
// different sentence.
type AmbiguousQuoteError struct {
	Quote   string
	Matches int
}

func (e *AmbiguousQuoteError) Error() string {
	return fmt.Sprintf("this quote occurs %d times in the document; send the text immediately "+
		"before it as prefix and/or after it as suffix so exactly one of them can be identified",
		e.Matches)
}

// EmptyQuoteError reports a quote that is blank once trimmed.
type EmptyQuoteError struct{}

func (e *EmptyQuoteError) Error() string { return "quote is empty" }

// QuoteTooLongError reports a quote past the anchor column's limit.
type QuoteTooLongError struct{ Bytes int }

func (e *QuoteTooLongError) Error() string {
	return fmt.Sprintf("quote is %d bytes; at most %d are stored. A quote is an identifier, not a "+
		"copy of the document — one sentence is enough", e.Bytes, maxQuoteBytes)
}

// ResolveQuote finds a quotation in a document's markdown and returns the anchor
// for it.
//
// prefix and suffix are optional and are used only to disambiguate: they are the
// text the caller saw immediately before and after the quote. They are not
// echoed back — the anchor's Prefix and Suffix are taken from the document at
// the match, because the anchor's job is to describe where the text is in this
// document, not to record what the caller believed. A caller that sent a slightly
// wrong neighbourhood still gets an anchor that will re-find itself later.
//
// It is refused rather than guessed at when the quote occurs more than once and
// the context does not settle it. See the note at the top of this file.
func ResolveQuote(source, quote, prefix, suffix string) (*Anchor, error) {
	exact := strings.TrimSpace(quote)
	if exact == "" {
		return nil, &EmptyQuoteError{}
	}
	if len(exact) > maxQuoteBytes {
		return nil, &QuoteTooLongError{Bytes: len(exact)}
	}

	candidates := verbatimOccurrences(source, exact)
	if len(candidates) == 0 && len(source) <= tolerantScanMaxSource {
		// The fallback for a quote taken from the rendered page: its characters
		// exist in the markdown with syntax interleaved.
		candidates = tolerantOccurrences(source, exact)
	}

	switch len(candidates) {
	case 0:
		return nil, &QuoteNotFoundError{Quote: exact}
	case 1:
		return anchorAt(source, exact, candidates[0]), nil
	}

	// Several. Let the context narrow them, exactly the way the frontend scores
	// them — but require it to leave one standing rather than picking a winner.
	if prefix != "" || suffix != "" {
		if best, unique := bestByContext(source, candidates, prefix, suffix); unique {
			return anchorAt(source, exact, best), nil
		}
	}
	return nil, &AmbiguousQuoteError{Quote: exact, Matches: len(candidates)}
}

// span is a half-open byte range in the source.
type span struct{ start, end int }

// verbatimOccurrences returns every literal occurrence of needle. Byte offsets
// fall out of it directly, which is why this is the exact path and the tolerant
// scan below is the approximate one.
func verbatimOccurrences(source, needle string) []span {
	var spans []span
	from := 0
	for {
		at := strings.Index(source[from:], needle)
		if at < 0 {
			break
		}
		start := from + at
		spans = append(spans, span{start, start + len(needle)})
		from = start + 1
		if len(spans) >= maxCandidates {
			break
		}
	}
	return spans
}

// tolerantOccurrences finds the needle allowing the source to carry markdown
// syntax the needle does not, and allowing whitespace runs to differ (a markdown
// soft line break renders as one space).
//
// Ported from locateQuoteInSource's fallback scan. It works in runes rather than
// UTF-16 code units, which agrees with the TypeScript for every character in the
// Basic Multilingual Plane — all prose, Cyrillic included. The two would count an
// astral-plane character (an emoji) differently, and the only thing that count
// feeds is maxSkipRun, a limit on how much markdown may sit between two letters
// of a quote; an emoji is not markdown syntax and does not appear in such a run.
func tolerantOccurrences(source, needle string) []span {
	if needle == "" {
		return nil
	}
	first, _ := utf8.DecodeRuneInString(needle)

	var spans []span
	for start, r := range source {
		// Only start where the first character could match — literally, or as the
		// head of a whitespace run.
		if r != first && (!unicode.IsSpace(first) || !unicode.IsSpace(r)) {
			continue
		}
		if end, ok := tolerantMatchAt(source, needle, start); ok {
			spans = append(spans, span{start, end})
			if len(spans) >= maxTolerantCandidates {
				break
			}
		}
	}
	return spans
}

// tolerantMatchAt tries to match needle starting exactly at BYTE offset from in
// source, and returns the byte offset one past the match.
//
// Byte offsets rather than an index into a pre-decoded []rune, deliberately: this
// is the one scan both directions of the primitive share. ResolveQuote sweeps it
// across the document, and SpanMatchesQuote runs it once at an offset a client
// supplied — which it can only do if a position in this function's vocabulary is
// the same thing as a position in the anchor column's. Decoding the whole body to
// address it would also make the guard pay 5 MiB of allocation to check one
// sentence.
func tolerantMatchAt(source, needle string, from int) (int, bool) {
	if from < 0 || from >= len(source) || !utf8.RuneStart(source[from]) {
		return 0, false
	}
	i, j := 0, from

	for i < len(needle) {
		if j >= len(source) {
			return 0, false
		}

		nc, nw := utf8.DecodeRuneInString(needle[i:])
		sc, sw := utf8.DecodeRuneInString(source[j:])
		if nc == sc {
			i += nw
			j += sw
			continue
		}

		// Whitespace is compared as runs: any run matches any run.
		if unicode.IsSpace(nc) && unicode.IsSpace(sc) {
			for i < len(needle) {
				r, w := utf8.DecodeRuneInString(needle[i:])
				if !unicode.IsSpace(r) {
					break
				}
				i += w
			}
			for j < len(source) {
				r, w := utf8.DecodeRuneInString(source[j:])
				if !unicode.IsSpace(r) {
					break
				}
				j += w
			}
			continue
		}

		// Skip a run of source-only markdown before giving up on this start.
		// Counted in runes, like the frontend counts it in UTF-16 units: the cap
		// is on how much markdown may sit between two letters of the quote, and a
		// byte count would make that budget depend on the alphabet.
		skipped := 0
		for j < len(source) && skipped < maxSkipRun {
			if tail := linkTail.FindString(source[j:]); tail != "" {
				j += len(tail)
				skipped += utf8.RuneCountInString(tail)
				continue
			}
			c := source[j]
			if c > 0x7f || !markupChars[c] {
				break
			}
			j++
			skipped++
		}
		if skipped == 0 {
			return 0, false
		}
	}

	return j, true
}

// SpanMatchesQuote reports whether [start, end) really is where quote sits in
// body — ResolveQuote read backwards, for an anchor a client computed itself.
//
// It exists because the anchor columns are documented as byte offsets into the
// markdown and nothing used to hold a writer to that. Two independently written
// clients put different units in them (PR #619 bytes, PR #621 character indices
// into the rendered text), and on ASCII the two are the same number, so both
// looked correct. On Cyrillic they differ by about half, and the row that results
// points confidently at a different sentence with no error anywhere.
//
// Three things it deliberately is NOT:
//
//   - not body[start:end] == quote. That equality is false for every legitimate
//     anchor whose selection crossed inline markup: the raw slice keeps the
//     markup (`**revert the image** first`) and the quote, taken from the
//     rendered text, does not (`revert the image first`). A strict-equality guard
//     would reject the correct client.
//   - not a UTF-8 validity check. A character offset almost always lands on a
//     valid rune boundary too — just not the right one — so a guard that only
//     looks at the encoding passes the exact bug it was written to catch. The
//     boundary test below is a cheap precondition, never the evidence.
//   - not a re-resolution. A quote occurring several times is fine here: the
//     caller told us which occurrence, and the question is only whether that
//     occurrence is the one they named.
func SpanMatchesQuote(body string, start, end int, quote string) bool {
	if quote == "" || start < 0 || end <= start || end > len(body) {
		return false
	}
	// Cheap precondition, not the check: a span that splits a character cannot be
	// a selection of one, and slicing on it would produce replacement characters
	// rather than the text.
	if !utf8.RuneStart(body[start]) || (end < len(body) && !utf8.RuneStart(body[end])) {
		return false
	}

	// A quote whose edges carry whitespace names the same range as its trimmed
	// form; ResolveQuote trims, so accepting both keeps the two directions
	// agreeing about one anchor.
	for _, candidate := range quoteForms(quote) {
		if body[start:end] == candidate {
			return true
		}
		if matchEnd, ok := tolerantMatchAt(body, candidate, start); ok && matchEnd == end {
			return true
		}
	}
	return false
}

// quoteForms is the quote as sent and, when they differ, as ResolveQuote would
// have stored it.
func quoteForms(quote string) []string {
	trimmed := strings.TrimSpace(quote)
	if trimmed == quote || trimmed == "" {
		return []string{quote}
	}
	return []string{quote, trimmed}
}

// bestByContext scores each candidate on how much of the supplied neighbourhood
// agrees with the document around it, and reports the best one — and whether it
// was the outright best. A tie is not a winner here, which is the difference
// between this and the frontend's bestCandidate.
func bestByContext(source string, candidates []span, prefix, suffix string) (span, bool) {
	best := candidates[0]
	bestScore := -1
	tied := false

	for _, c := range candidates {
		score := contextScore(source, c, prefix, suffix)
		switch {
		case score > bestScore:
			best, bestScore, tied = c, score, false
		case score == bestScore:
			tied = true
		}
	}
	// A best score of zero means "the quote is here but nothing around it agrees"
	// for every candidate, which is not a match on context at all.
	return best, !tied && bestScore > 0
}

// contextScore is how well the document around a candidate agrees with the
// remembered neighbourhood: the longest common suffix of what precedes it with
// prefix, plus the longest common prefix of what follows it with suffix. Same
// measure as contextScore in anchor.ts.
func contextScore(source string, c span, prefix, suffix string) int {
	before := runesBefore(source, c.start, len([]rune(prefix)))
	after := runesAfter(source, c.end, len([]rune(suffix)))
	return commonSuffixLen(before, prefix) + commonPrefixLen(after, suffix)
}

// runesBefore returns up to n runes of source ending at byte offset at.
func runesBefore(source string, at, n int) string {
	if n <= 0 || at <= 0 {
		return ""
	}
	head := []rune(source[:at])
	if len(head) > n {
		head = head[len(head)-n:]
	}
	return string(head)
}

// runesAfter returns up to n runes of source starting at byte offset at.
func runesAfter(source string, at, n int) string {
	if n <= 0 || at >= len(source) {
		return ""
	}
	tail := []rune(source[at:])
	if len(tail) > n {
		tail = tail[:n]
	}
	return string(tail)
}

func commonSuffixLen(a, b string) int {
	ra, rb := []rune(a), []rune(b)
	n := 0
	for n < len(ra) && n < len(rb) && ra[len(ra)-1-n] == rb[len(rb)-1-n] {
		n++
	}
	return n
}

func commonPrefixLen(a, b string) int {
	ra, rb := []rune(a), []rune(b)
	n := 0
	for n < len(ra) && n < len(rb) && ra[n] == rb[n] {
		n++
	}
	return n
}

// anchorAt builds the anchor for a located span, taking the neighbourhood from
// the document rather than from the caller.
func anchorAt(source, exact string, at span) *Anchor {
	return &Anchor{
		Exact:  exact,
		Prefix: runesBefore(source, at.start, contextLength),
		Suffix: runesAfter(source, at.end, contextLength),
		Start:  at.start,
		End:    at.end,
	}
}
